# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: browser/base/content/test/tabs/browser_close_during_beforeunload.js
# Commit: baedd4ec9add
# Full Hash: baedd4ec9add5edf5d78168eb46b570fb1edcaf2
# Author: Kris Maglione <maglione.k@gmail.com>
# Date: 2020-10-06 04:10:51
# Description:
#   Bug 1667485: Fix inifinite nested event loop spinning in corner cases. r=nika
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D91886
# ==============================================================================

diff -r 9e904ebcedef -r baedd4ec9add browser/base/content/test/tabs/browser_close_during_beforeunload.js
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/browser/base/content/test/tabs/browser_close_during_beforeunload.js	Mon Oct 05 17:57:48 2020 +0000
@@ -0,0 +1,36 @@
+"use strict";
+
+// Tests that a second attempt to close a window while blocked on a
+// beforeunload confirmation ignores the beforeunload listener and
+// unblocks the original close call.
+
+const DIALOG_TOPIC = "tabmodal-dialog-loaded";
+
+add_task(async function() {
+  await SpecialPowers.pushPrefEnv({
+    set: [["dom.require_user_interaction_for_beforeunload", false]],
+  });
+
+  let win = await BrowserTestUtils.openNewBrowserWindow();
+
+  let browser = win.gBrowser.selectedBrowser;
+  await BrowserTestUtils.loadURI(browser, "http://example.com/");
+
+  await SpecialPowers.spawn(browser, [], () => {
+    // eslint-disable-next-line mozilla/balanced-listeners
+    content.addEventListener("beforeunload", event => {
+      event.preventDefault();
+    });
+  });
+
+  let confirmationShown = false;
+
+  BrowserUtils.promiseObserved(DIALOG_TOPIC).then(() => {
+    confirmationShown = true;
+    win.close();
+  });
+
+  win.close();
+  ok(confirmationShown, "Before unload confirmation should have been shown");
+  ok(win.closed, "Window should have been closed after second close() call");
+});